htpy/utils/html_to_htpy.py

```diff
@@ -0,0 +1,208 @@
+                if not val:
+                    _attrs += f"{key}=True,"
+
+                else:
+                    _attrs += f'{key}="{val}",'
+
+        if _positional_attrs or _kwattrs:
+            _attrs = _attrs[:-1] + ")"
+
+        _children: str = ""
+        if self.children:
+            _children += "["
+            for c in self.children:
+                if isinstance(c, Tag):
+                    _children += c.serialize(shorthand_id_class=shorthand_id_class)
+                else:
+                    _children += str(c)
+
+                _children += ","
+
+            _children = _children[:-1] + "]"
+
+        return f"{_type}{_attrs}{_children}"
+
+
+class HTPYParser(HTMLParser):
+    def __init__(self):
+        self._collected: list[Tag | str] = []
+        self._current: Tag | None = None
+        super().__init__()
+
+    def handle_starttag(self, tag: str, attrs: list[tuple[str, str | None]]):
+        t = Tag(tag, attrs, parent=self._current)
+
+        if not self._current:
+            self._collected.append(t)
+        else:
+            self._current.children.append(t)
+
+        self._current = t
+
+    def handle_startendtag(self, tag: str, attrs: list[tuple[str, str | None]]):
+        t = Tag(tag, attrs, parent=self._current)
+
+        if not self._current:
+            self._collected.append(t)
+        else:
+            self._current.children.append(t)
+
+    def handle_endtag(self, tag: str):
+        if not self._current:
+            raise Exception(
+                f"Error parsing html: Closing tag {tag} when not inside any other tag"
+            )
+
+        if not self._current.type == tag:
+            raise Exception(
+                f"Error parsing html: Closing tag {tag} does not match the currently open tag ({self._current.type})"
+            )
+
+        self._current = self._current.parent
+
+    def handle_data(self, data: str):
+        if not data.isspace():
+            stringified_data = _convert_data_to_string(data)
+
+            if self._current:
+                self._current.children.append(stringified_data)
+            else:
+                self._collected.append(stringified_data)
+
+    def serialize_python(self, shorthand_id_class: bool = False, format: bool = False):
+        o = ""
+
+        if len(self._collected) == 1:
+            o += _serialize(self._collected[0], shorthand_id_class)
+
+        else:
+            o += "["
+            for t in self._collected:
+                o += _serialize(t, shorthand_id_class) + ","
+            o = o[:-1] + "]"
+
+        if format:
+            try:
+                import black
+            except:
+                raise Exception(
+                    "Cannot import formatter. Please ensure black is installed."
+                )
+
+            return black.format_str(
+                o, mode=black.FileMode(line_length=80, magic_trailing_comma=False)
+            )
+        else:
+            return o
```

Comment:
I was looking into this a bit, and decided to drop the formatting stuff for now. I do find it interesting though!

A couple of notes. Disclaimer: I have not personally started using ruff yet, although I probably should. So I don't know ins and outs.

1. I did not find a python interface for ruff, so I'm not sure it can be done in the same way as the black one was done. Maybe it can be solved in some other way though.
2. It seems from the docs that it is only able to format files and cannot (?) receive a string/input stream

Reply:
Feel free to drop the formatting altogether now! It is better to merge this as is and then we can add more features later.

Using a subprocess may be more straightforward to support both. It is enough just to have black or ruff on PATH to make it work:

```pycon
Python 3.12.3 (main, Apr  9 2024, 08:09:14) [Clang 16.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> subprocess.run(['black', '-q', '-'], input='def foo(a   = 1):pass'.encode('utf8')).stdout
def foo(a=1):
    pass
>>> subprocess.run(['ruff', 'format', '-'], input='def foo(a   = 1):pass'.encode('utf8')).stdout
def foo(a=1):
    pass
>>>
```

Reply:
I added this in the latest commit.